include/mq/api/MacroDataTypes.h

```diff
@@ -0,0 +1,552 @@
+/*
+ * MacroQuest: The extension platform for EverQuest
+ * Copyright (C) 2002-2023 MacroQuest Authors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License, version 2, as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#pragma once
+
+#include "mq/base/Common.h"
+#include "mq/base/Deprecation.h"
+
+#include "eqlib/base/Color.h"
+#include "eqlib/CXStr.h"
+#include "eqlib/Items.h"
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string_view>
+#include <variant>
+
+#define TypeMember(name) AddMember(static_cast<int>(name), #name)
+#define ScopedTypeMember(scope, name) AddMember(static_cast<int>(scope::name), #name)
+#define TypeMethod(name) AddMethod(static_cast<int>(name), #name)
+#define ScopedTypeMethod(scope, name) AddMethod(static_cast<int>(scope::name), #name)
+
+
+namespace mq {
+
+namespace datatypes {
+	class MQ2Type;
+}
+using datatypes::MQ2Type;
+struct MQTypeVar;
+
+
+namespace detail {
+
+	template <typename T>
+	constexpr bool is_shared_ptr = false;
+
+	template <typename T>
+	constexpr bool is_shared_ptr<std::shared_ptr<T>> = true;
+
+	template <typename T>
+	struct shared_ptr_element_type { using type = T; };
+
+	template <typename T>
+	struct shared_ptr_element_type<std::shared_ptr<T>> { using type = typename std::shared_ptr<T>::element_type;  };
+
+	template <typename From, typename To>
+	static void ConvertData(const From& input, std::optional<To>& output)
+	{
+		if constexpr (std::is_convertible_v<From, To>)
+			output = static_cast<To>(input);
+		else
+		{
+			WriteChatf("Tried to convert unlike types %s and %s", type_name<From>().c_str(), type_name<To>().c_str());
+
+			if (gMacroBlock != nullptr && gMacroBlock->Line.find(gMacroBlock->CurrIndex) != gMacroBlock->Line.end())
```

Comment:
MQ2Globals.h needs to be included for this to make sense, should that just be included in this header?

Reply:
by "this" I mean using WriteChatf and gMacroStack